Tests/PromiseTests.cs

```diff
@@ -1472,5 +1472,65 @@ public void rejected_reject_callback_is_caught_by_chained_catch()
 
             Assert.Equal(expectedException, actualException);
         }
+
+        [Fact]
+        public void rejected_catch_returning_resolved_promise_state_is_adopted()
+        {
+            var promise = new Promise<int>();
+            int expectedValue = 1;
+            int actualValue = 0;
+
+            promise.Catch(err => Promise<int>.Resolved(expectedValue))
+                   .Then(result => { actualValue = result; });
+
+            promise.Reject(new Exception());
+
+            Assert.Equal(expectedValue, actualValue);
+        }
+
+        [Fact]
+        public void rejected_catch_returning_rejected_promise_state_is_adopted()
+        {
+            var promise = new Promise<int>();
+            int expectedValue = 1;
+            int actualValue = 0;
+
+            promise.Catch(err => Promise<int>.Rejected(new Exception()))
+                   .Catch(err => actualValue = expectedValue);
+
+            promise.Reject(new Exception());
+
+            Assert.Equal(expectedValue, actualValue);
+        }
+
+        [Fact]
+        public void rejected_catch_returning_resolved_promise_state_is_adopted_nongeneric()
+        {
+            var promise = new Promise();
+            int expectedValue = 1;
+            int actualValue = 0;
+
+            promise.Catch(err => Promise.Resolved())
+                   .Then(() => { actualValue = expectedValue; });
+
+            promise.Reject(new Exception());
+
+            Assert.Equal(expectedValue, actualValue);
+        }
+
+        [Fact]
+        public void rejected_catch_returning_rejected_promise_state_is_adopted_nongeneric()
+        {
+            var promise = new Promise();
+            int expectedValue = 1;
+            int actualValue = 0;
+
+            promise.Catch(err => Promise.Rejected(new Exception()))
+                   .Catch(err => actualValue = expectedValue);
+
+            promise.Reject(new Exception());
+
+            Assert.Equal(expectedValue, actualValue);
+        }
```

Comment:
You should add some tests to make sure the value passed into onResolved and onRejected are the same as the promise returned in the first catch.